[inductor] Refactor generate_kernel_call

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102093
* #102083
* #102019
* __->__ #102018
* #102017

Summary: Refactor generate_kernel_call to support codegen call to Triton
kernel

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10

desertfire committed:
[inductor] Refactor generate_kernel_call

Summary: Refactor generate_kernel_call to support codegen call to Triton
kernel

[ghstack-poisoned]

ngimel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -649,8 +649,20 @@ def generate_profiler_mark_wrapper_call(self, stack):
-    def generate_kernel_call(self, name, call_args, device_index=None, cpp=False):
-        self.writeline(self.wrap_kernel_call(name, call_args))
+    def generate_kernel_call(
+        self, name, call_args, grid=None, device_index=None, cuda=None
```
what type is `cuda`? You are passing `cuda=True`, so default should probably be `False`, not `None`?

desertfire replied:
I will change it to `cuda=True` as the triton backend is the default one.

ngimel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1124,11 +1136,10 @@ def write_prefix(self):
+        self, name: str, kernel: str, metadata: Optional[str] = None, cuda=True
     ):
-        if cpp:
+        if not cuda:
```
what's the reason for changing arg from cpp to cuda?

desertfire replied:
The main reason is because `cpp` means two things in inductor's codegen, the cpu backend and the cpp wrapper codegen. Because I added generate_kernel_call for the triton backend into the base class here, WrapperCodeGen, now we need a flag to distinguish the two backends, and `cpp` wouldn't be a good naming here. `cuda` may not be the best naming, but it's slightly better.

EikanWang replied:
@desertfire , I'm trying to enable Inductor to support other backends at runtime https://github.com/pytorch/pytorch/pull/100706. The design assumes the backend just needs to customize two meta classes, `BaseScheduling` and `wrapper.WrapperCodeGen`, by overriding some member functions. Regarding the `define_kernel`, I'm kind of confused by the `cuda` parameter because it is the logic of  `CudaWrapperCodeGen` and it has implied `cuda` is `True`. My question is this parameter is must-have.

desertfire replied:
Initially I didn't have `cuda` as a parameter, but the problem appears when we have a mix of devices (see `test_multi_device`). Because we only keep one wrapper codegen instance, i.e. CudaWrapperCodeGen when there is a mix of cuda and cpu tensors, we need this additional parameter to help generate correct kernel definition and kernel call.

desertfire committed:
Update on "[inductor] Refactor generate_kernel_call"


Summary: Refactor generate_kernel_call to support codegen call to Triton
kernel

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge
